chore: fix compiler warnings

mwerle opened the pull request:
This PR fixes all warnings (mostly deprecations) in the codebase when building against QT5.15 using both `clang` and `gcc`.

It also ensures the codebase can still be compiled with versions of QT prior to 5.14 via a new compatibility header file `qtsupport.h` and, where that was not possible, via alternate code compilation paths via a #if..else..endif. The latter approach was required for the `QSet` initializer.

Out-of-scope:
* deprecations/warnings in the dependencies

Changes:
* add default to switch statements throwing an exception on unhandled case. This removes `control reaches end of non-void function` in many functions.
* fix QT5.14 deprecations, which will also make it easier to move to QT6
  - fix QSet initialization
  - rename `endl` -> `Qt::endl`
  - rename `QString::SkipEmptyParts` -> `Qt::SkipEmptyParts`
  - and more...
* Scintilla editor 
  - add missing `override`
  - replace deprecated functions
* Settings
  - fix deprecated QDir assignment
  - code-style: replace assignment with initializers (to match QDir in this file)

NOTE: did not fix the deprecated usage of QProcess, as that requires a bit more throught as to how we should handle the list of configured commands and their arguments.

Testing:
* The unit tests pass.
* anecdotal (ie, own usage) testing showed no issues.

Murmele commented:
Thanks, I will have a look into it.

mwerle commented:
G'day, apologies for the failed formatting; I didn't notice the script..

Is there a way I can run the CI tasks locally prior to sending a PR? I did also try to trigger them via GitHub after I created the PR but couldn't figure out how.

Murmele commented:
> G'day, apologies for the failed formatting; I didn't notice the script..
> 
> Is there a way I can run the CI tasks locally prior to sending a PR? I did also try to trigger them via GitHub after I created the PR but couldn't figure out how.

You can run the unittests by entering the build folder and execute `ninja test` or `ctest`

You can use the format script to format the code and then comitting it

mwerle commented:
> > G'day, apologies for the failed formatting; I didn't notice the script..
> > Is there a way I can run the CI tasks locally prior to sending a PR? I did also try to trigger them via GitHub after I created the PR but couldn't figure out how.
> 
> You can run the unittests by entering the build folder and execute `ninja test` or `ctest`
> 
> You can use the format script to format the code and then comitting it

Sure, but the CI does more such as performing builds for a variety of platforms and creating the Flatpak etc.

Hmm, formatting is still failing? But I ran clang-format locally and committed all the requested changes and re-running it does not show up any new deviations.

Murmele commented:
This is because of the clang-format version which is 13, it makes problems with clang-format 14 ...

mwerle commented:
> This is because of the clang-format version which is 13, it makes problems with clang-format 14 ...

Right, well, your `cl-format.sh` always looked for the latest version. I've modified it to ensure it uses v13. Also fixed the cmake-format to use a python venv; IMHO poor practise to install stuff into a developers' machine globally.

Finally, added a configuration to run clang-format v13 as a pre-commit hook. "Works For Me" but might require additional testing on a variety of build environments. It's optional to use in either case.

Murmele commented on src/app/CustomTheme.cpp:
```diff
@@ -435,6 +435,9 @@ QColor CustomTheme::commitEditor(CommitEditor color) {
       return commitEditor.value("lengthwarning").value<QColor>();
+    default:
+      throw std::runtime_error("Not Implemented or invalid enum " +
+                               std::to_string(static_cast<int>(color)));
```
I updated it, what do you think? So we are checking during compile time and not crashing during runtime

mwerle replied:
> I updated it, what do you think? So we are checking during compile time and not crashing during runtime

Sorry, I was away over the weekend.  

Yes, I much prefer compile-time-checks to runtime checks for non-security-critical code.  I hadn't seen the `-Wswitch-enum` before, and it seems it's not part of `-Wall`..

OTOH, it seems it's not a bullet-proof solution either depending on the compiler used:
https://stackoverflow.com/questions/58631852/handling-of-switch-enum-class-returns-in-clang-gcc-and-icc-consistently

And, of course, it doesn't protect at runtime if somewhere in the code, some code-path ends up passing an illegal value into the switch by casting a random integer to the enum (something I thought C++ protected against, but it appears it doesn't), at which point you end up with a hard-to-identity bug.

Ultimately your call how you'd like to resolve this, unfortunately it seems there's no single "golden" answer.

Murmele commented:
[default.patch](https://github.com/Murmele/Gittyup/files/12455726/default.patch)

mwerle commented:
So after looking into it a it more, I must say I'm not keen on the `-Werror=switch-enum` switch.  It forces enumerating all items in a switch statement even if there is a `default`. IMHO this just adds noise to the code and makes it less readable.

As far as I see it, there are basically 2 situations:
* developer wants to handle every case explicitly, meaning there should not be a default. This allows the compiler to warn if a case was forgotten, or during later development when a new case is added to ensure it is handled in all affected switch statements. This is primarily when the developer controls both the enum as well as the logic.
* developer only wants to handle a subset of cases. In this case the developer should not have to list all the cases which are of no interest but instead use default. This is often the case when the enum comes from a third-party source such as an external library.

So I've changed the switch to `-Werror=switch`, and I've also moved all my `throw` statements outside the switch statements; in retrospect having them in the `default` clause was a poor choice by myself initially.  This also means that I did not apply your patch and, in fact, deleted several items from some switch statements and replaced them with a `default` entry instead.

Please let me know if some/all of your patch should be applied after all and I'll go through it again.

----

I've also briefly enabled `-Wall` and fixed a few more warnings. `-Wextra` was just too much for now.  Ultimately I would encourage the project to enable all warnings and aim towards a zero-warning compilation policy. While many warnings are advisory (and sometimes a right-royal pain to remove for questionable benefit), a few are actually indicative of real code problems. It's difficult to find a new problem if it is swamped by dozens/hundreds of innocuous warnings.

Murmele commented:
Hi @mwerle sorry that I did not apply my patch my self, but I was not able to do from the codespace for some reason.
Yes I agree with you, sometimes Werror=switch-enum is to strict, but for me the default is like a joker and if you add a new enum you might forgett to implement somewhere the case. But neverthless right now we can let it as it is, thank you for your great work!

mwerle commented:
No worries re the patch. I'm sorry, I didn't go through it in detail, but mostly it seemed to be about adding a `default` case to many of the switch statements. As I argued, adding a default actually prevents the compiler from warning about missing options (unless you use the -Wswitch-enum flag).

If you'd still like me to apply it, I'll of course do so. And apologies for taking up so much of your time on a "useless" refactor.. hoping to contribute some more useful changes in the future :)

Murmele commented on src/editor/TextEditor.cpp:
```diff
@@ -43,8 +43,6 @@ QPixmap stagedUnstagedIcon(const bool &checked, const QColor &background,
       QRect(QPoint(0, 0), QSize(fontHeight - 2, fontHeight - 2)));
 }
 
-const float textHeightFactorCheckBoxSize = 2.0;
```
Use instead 
```cpp
#if defined(FLATPAK)
  const float textHeightFactorCheckBoxSize = 2.0;
#endif
```

mwerle replied:
Thanks; I guess I'll need to figure out how to build flatpaks if there's going to be more differences like this..

mwerle commented:
Hmm.... for some reason my git pre-commit hook isn't being invoked; at least, not when using `gittyup`. I did notice it being invoked when using the command-line a few days ago.  Something to investigate further.

Murmele commented:
> 

No problem, you can fix it in another MR. I will merge it as soon as the CI is finished. Thank you!
